### PR TITLE
Refactored cuda kernel cache

### PR DESCRIPTION
**Pull request by jjsjann123**
1. Cleaned up `CudaKernel`; (We still have the block/grid thing there used in `runTestKernel`, they need to go at some point. I'll work it out with Kevin at a later time).
2. Updated kernel selection in cache to utilize `LaunchConfigType::Compatible`;
3. Updated graph stringify to preserve rank of tensor (removed the unnecessary `EraseShapeInformation(graph)`).

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/fusion.h [hunk 1]
```diff
@@ -63,7 +63,7 @@ class TORCH_CUDA_API FusionGuard {
 
   // Set the active fusion so it can be manipulated.
   FusionGuard(Fusion* fusion);
-  FusionGuard(const cuda::CudaKernel* cuda_kernel);
+  FusionGuard(cuda::CudaKernel* cuda_kernel);
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/fusion.h]: updating signature to be consistent with `FusionGuard(Fusion* fusion)`

File: torch/csrc/jit/codegen/cuda/kernel.cpp [hunk 2]
```diff
@@ -264,40 +222,40 @@ bool validateKernelArg(
 }
 
 void validateKernelArgs(
-    const CudaKernel& entry,
+    CudaKernel* entry,
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/kernel.cpp:226]: @naoyam I'm removing the `const` here as we don't have the right const model on `FusionGuard`
Feel free to yell at me if there's a better way to do this.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/kernel.cpp:226]: Sounds reasonable to me.

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 3]
```diff
@@ -9,19 +11,39 @@ namespace fuser {
 namespace cuda {
 
 at::optional<CudaKernel*> CudaKernelCache::getKernelPtr(
-    const at::ArrayRef<c10::IValue> inputs) {
-  for (auto& iter : kernels_) {
-    if (iter.first->matchKernelSize(inputs)) {
-      return &(iter.second);
+    const at::ArrayRef<c10::IValue> inputs,
+    const std::vector<int64_t>& broadcasted_shape) {
+  for (auto& cuda_kernel : kernels_) {
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp:16]: We iterate through all kernels and check if a given kernel could be used to run on the set of inputs.

TODO: we can cache the kernel selection for a set of input shapes, so we are not evaluating the flag every iteration. (maybe maintain a fixed-size table for kernel selection keys on input sizes)

File: test/cpp/jit/test_gpu.cpp [hunk 4]
```diff
@@ -410,7 +411,7 @@ void testGPU_FusionClear() {
     tv1->computeAt(tv3, -1);
   }
 
-  prog.device_ = 0;
+  prog.getDevice() = 0;
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp]: if you can set using this interface, please just use `prog.device()` if you really want it to be a method, otherwise use `getDevice()` and `setDevice(...)`

File: test/cpp/jit/test_gpu.cpp [hunk 5]
```diff
@@ -1017,18 +1017,18 @@ void testGPU_FusionParser() {
   }
 
   torch::jit::fuser::cuda::CudaKernel prog;
-  Fusion& fusion = *prog.fusion_;
+  prog.setFusionPtr(fuser::cuda::parseJitIR(g));
+  Fusion& fusion = *prog.getFusionPtr();
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp]: can't we just work with `Fusion*` instead of dereferencing?

File: test/cpp/jit/test_gpu.cpp [hunk 6]
```diff
@@ -1017,18 +1017,18 @@ void testGPU_FusionParser() {
...
   prog.grid(4);
   prog.block(32);
-  prog.device_ = 0;
-  fuser::cuda::parseJitIR(g, &prog);
+  prog.getDevice() = 0;
   auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
   at::Tensor input1 = at::randn({16}, options);
   at::Tensor input2 = at::randn({16}, options);
-  fuser::cuda::scheduleFusion(prog.fusion_.get(), {input1, input2});
+  fuser::cuda::scheduleFusion(prog.getFusionPtr(), {input1, input2});
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp]: Just use `fusion`?

File: test/cpp/jit/test_gpu.cpp (hunk 5 above)
Comment (csarofeen) [reply test/cpp/jit/test_gpu.cpp]: Applies moving forward.

File: test/cpp/jit/test_gpu.cpp (hunk 6 above)
Comment (csarofeen) [reply test/cpp/jit/test_gpu.cpp]: Applies moving forward.

**Review by csarofeen — APPROVED** [review]
Nit in test, otherwise LGTM.

File: test/cpp/jit/test_gpu.cpp (hunk 4 above)
Comment (tlemo) [reply test/cpp/jit/test_gpu.cpp]: I agree, no need for the "get" prefix. Even when we have setters, we can still use the plain name for getters (ex. device() + setDevice())

File: torch/csrc/jit/codegen/cuda/fusion.h (hunk 1 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/fusion.h]: since you touched this, please make both constructors `explicit`

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 7]
```diff
@@ -9,19 +11,39 @@ namespace fuser {
...
   return at::nullopt;
 }
 
 CudaKernel* CudaKernelCache::allocateKernelInCache(
-    std::unique_ptr<KernelArgsReq>&& args_req) {
-  kernels_.emplace_back(std::move(args_req), CudaKernel());
-  return &(kernels_.back().second);
+    const at::ArrayRef<c10::IValue> inputs) {
+  kernels_.emplace_back();
+  return &(kernels_.back());
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp]: nit: we don't need the extra parentheses

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 8]
```diff
@@ -9,19 +11,39 @@ namespace fuser {
...
   }
   return at::nullopt;
 }
 
 CudaKernel* CudaKernelCache::allocateKernelInCache(
-    std::unique_ptr<KernelArgsReq>&& args_req) {
-  kernels_.emplace_back(std::move(args_req), CudaKernel());
-  return &(kernels_.back().second);
+    const at::ArrayRef<c10::IValue> inputs) {
+  kernels_.emplace_back();
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp:45]: emplace back what?

File: torch/csrc/jit/codegen/cuda/kernel_cache.h [hunk 9]
```diff
@@ -13,21 +13,125 @@ namespace jit {
...
+      const at::Tensor& val,
+      c10::optional<at::IntArrayRef> broadcasted_size = c10::nullopt) {
+    if (broadcasted_size) {
+      // [Note - broadcast support in integration]
+      // PyTorch follows numpy broadcasting rule.
+      // (https://numpy.org/doc/stable/user/basics.broadcasting.html)
+      //
+      // So in case where the rank of two operators differ, we align them on
+      // the higher dimensions, hence the offset o_dim-b_dim to the index here.
+      int b_dim = (int)broadcasted_size->size();
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.h]: nit: functional style cast instead of c-style casts

`int(broadcasted_size->size())`

File: torch/csrc/jit/codegen/cuda/kernel_cache.h [hunk 10]
```diff
@@ -13,21 +13,125 @@ namespace jit {
...
+      int o_dim = (int)val.dim();
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.h]: const?

File: torch/csrc/jit/codegen/cuda/kernel_cache.h [hunk 11]
```diff
@@ -13,21 +13,125 @@ namespace jit {
...
+      TORCH_CHECK(b_dim >= o_dim);
+      for (int i = 0; i < b_dim; i++) {
+        sizes.push_back(broadcasted_size->at(i));
+        int index = i + o_dim - b_dim;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.h]: const?

File: torch/csrc/jit/codegen/cuda/kernel_cache.h [hunk 12]
```diff
@@ -13,21 +13,125 @@ namespace jit {
...
+          strides.push_back(val.strides()[index]);
+        } else {
+          TORCH_CHECK(
+              val.sizes()[index] == 1,
+              "Not compatible dimension size for broadcast");
+          strides.push_back(0);
+        }
+      }
+    } else {
+      auto o_dim = val.dim();
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.h]: const?

File: torch/csrc/jit/codegen/cuda/kernel_cache.h [hunk 13]
```diff
@@ -13,21 +13,125 @@ namespace jit {
...
+
+  Fusion* getFusionPtr() {
+    return fusion_.get();
+  }
+
+  const Fusion* getFusionPtr() const {
+    return fusion_.get();
+  }
+
+  CUmodule& getModule() {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.h]: 1. what is the convention used here? return pointers or return references? I think we should pick one for consistency

2. is returning mutable references intentional?

File: torch/csrc/jit/codegen/cuda/kernel_cache.h [hunk 14]
```diff
@@ -13,21 +13,125 @@ namespace jit {
...
+    return module_;
+  }
+
+  CUfunction& getFunction() {
+    return function_;
+  }
+
+  int16_t& getDevice() {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.h]: why return a mutable reference - do we want to allow changing it?

File: torch/csrc/jit/codegen/cuda/kernel_cache.h [hunk 15]
```diff
@@ -13,21 +13,125 @@ namespace jit {
...
+    return device_;
+  }
+
+  const int16_t& getDevice() const {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.h]: why reference?

File: torch/csrc/jit/codegen/cuda/kernel_cache.h [hunk 16]
```diff
@@ -13,21 +13,125 @@ namespace jit {
...
+    return device_;
+  }
+
+  bool hasRNG() {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.h]: ideally this should be const method. also, do we really need the fusion guard?

File: torch/csrc/jit/codegen/cuda/kernel_cache.h [hunk 17]
```diff
@@ -13,21 +13,125 @@ namespace jit {
...
+  int16_t device_;
+  CUmodule module_;
+  CUfunction function_;
+  std::unique_ptr<Fusion> fusion_;
+
+  // WARNING:
+  // Block and Grid dimension setting is here for testing purposes only
+  // These are not here for general use and only for use with
+  // the runTestKernel() function.
+ public:
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.h:108]: 1. public, then private then public makes it hard to read (ex. figure out the public interface)
  -> move the public functions to the public section above?

2. do block_ and grid_ also have to be public?

3. if block() and grid() are only used by runTestKernel() we may have better options:
a. private + friend runTestKernel() - easy option
b. create a testing interface (so we don't have to friend eveything in this class) - may not be worth it here

File: torch/csrc/jit/codegen/cuda/manager.cpp [hunk 18]
```diff
@@ -86,42 +76,42 @@ class CudaFusionManager {
...
-      //   2. launch config;
-      parseJitIR(graph, cuda_kernel.value());
-      scheduleFusion(cuda_kernel.value()->fusion_.get(), inputs);
+      cuda_kernel->setFusionPtr(parseJitIR(graph));
+      TORCH_INTERNAL_ASSERT(
+          cuda_kernel->getFusionPtr() != nullptr,
+          "parser failed to construct a fusion from PyTorch JIT graph\n");
+
+      // magic scheduler
+      scheduleFusion(cuda_kernel->getFusionPtr(), inputs);
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/manager.cpp]: add a comment reminding that the fusion will be mutated here?
(I'd actually prefer a different interface, where scheduleFusion() returns the new Fusion, but that may be outside the scope of this PR)

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp (hunk 8 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/kernel_cache.cpp:45]: emplace_back with empty argument, so pushing back an element with default constructor. I can be more explicit and use `kernels_.emplace_back(CudaKernel())`.

I've been skipping the argument quite often just to get shorter code, is it a bad style?

File: torch/csrc/jit/codegen/cuda/kernel_cache.h (hunk 13 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/kernel_cache.h]: 1. I'll update to `module()`.
2. mutable reference is intentional. Alternatively I can switch to returning pointer.

File: torch/csrc/jit/codegen/cuda/kernel_cache.h (hunk 14 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/kernel_cache.h]: Yes, there is a place to update it.

I'm using return by reference so we can have consistent way to update/access `device_`/`function_`/`module_`. It looks still messy right now. Let me try again.

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp (hunk 8 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/kernel_cache.cpp:45]: I wanted to make sure it's intentional.

Personally I'd prefer the more explicit push_back(CudaKernel()), but that's a matter of taste. Note that emplace_back(CudaKernel()) doesn't really buy you anything compared to push_back() - both will use the copy constructor.

File: torch/csrc/jit/codegen/cuda/kernel_cache.h (hunk 14 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/kernel_cache.h]: Just an idea: it may be less surprising to use an explicit setter here: setDevice()

File: torch/csrc/jit/codegen/cuda/manager.cpp (hunk 18 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/manager.cpp]: Agreed.

File: torch/csrc/jit/codegen/cuda/kernel_cache.h (hunk 16 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/kernel_cache.h]: At some point in time, we need it when traversing the graph. I'm not sure if it's still true but I'm also too scared to remove it. :fearful: 

I'm marking this a const method.

File: torch/csrc/jit/codegen/cuda/kernel_cache.h (hunk 17 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/kernel_cache.h:108]: 1. block / grid is only used by testing. I'm hoping that they will go away in a few round of PRs as Kevin and I are merging code path. Hence the intention to leave them in the corner.
2. I think so, `block_` and `grid_` are used as accessors.
3. both option sounds reasonable.

Would you mind if I leave it as-is and do it in a following PR to merge `runKernel` and `runTestKernel`? Hence this wouldn't be needed any more.

File: torch/csrc/jit/codegen/cuda/kernel_cache.h (hunk 14 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/kernel_cache.h]: Missed this comment. but I happen to use `setDevice()` anyway.
Great minds think alike, so do us. :laughing:

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp (hunk 8 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/kernel_cache.cpp:45]: > emplace_back(CudaKernel()) doesn't really buy you anything compared to push_back() - both will use the copy constructor.

Agree that they behave the same in this case and they both use move constructor, but that's not saving anything.

emplace_back() is different though, as you construct in-place and there's not copy.

**Review by tlemo — APPROVED** [review]
